benchmark/single-source/FindStringNaive.swift

```diff
@@ -0,0 +1,185 @@
+  _ string: S,
+  needle: Character
+) -> String.Index? {
+  var index = string.startIndex
+  while index < string.endIndex {
+    let nextIndex = string.index(after: index)
+    if string[index] == needle {
+      return index
+    }
+    index = nextIndex
+  }
+  return nil
+}
+
+func findThreeRecursive<S: StringProtocol>(
+  _ string: S,
+  needle1: Character,
+  needle2: Character?,
+  needle3: Character?
+) -> String.Index? {
+  var index = string.startIndex
+  while index < string.endIndex {
+    let nextIndex = string.index(after: index)
+    if string[index] == needle1 {
+      // Check subsequent needles recursively (if applicable)
+      guard let needle2 = needle2 else { return index }
+
+      if findThreeRecursive(
+        string[nextIndex...].prefix(2), needle1: needle2, needle2: needle3, needle3: nil
+      ) == nextIndex {
+        return index
+      }
+    }
+    index = nextIndex
+  }
+  return nil
+}
+
+func findOneOnUTF8Collection<Bytes: Collection>(
+  _ string: Bytes,
+  needle: UInt8
+) -> Bytes.Index? where Bytes.Element == UInt8 {
+  var index = string.startIndex
+  while index < string.endIndex {
+    let nextIndex = string.index(after: index)
+    if string[index] == needle {
+      return index
+    }
+    index = nextIndex
+  }
+  return nil
+}
+
+func findThreeOnUTF8Collection<Bytes: Collection>(
+  _ string: Bytes,
+  needle1: UInt8,
+  needle2: UInt8?,
+  needle3: UInt8?
+) -> Bytes.Index? where Bytes.Element == UInt8 {
+  var index = string.startIndex
+  while index < string.endIndex {
+    let nextIndex = string.index(after: index)
+    if string[index] == needle1 {
+      // Check subsequent needles recursively (if applicable)
+      guard let needle2 = needle2 else { return index }
+
+      if findThreeOnUTF8Collection(
+        string[nextIndex...].prefix(2), needle1: needle2, needle2: needle3, needle3: nil
+      ) == nextIndex {
+        return index
+      }
+    }
+    index = nextIndex
+  }
+  return nil
+}
+
+@inline(never)
+func runBenchLoop1Substring(iterations: Int) {
+  for _ in 0 ..< iterations {
+    precondition(findOne(longStringFoFoFoFox![...], needle: "x") != nil)
+  }
+}
+
+@inline(never)
+func runBenchLoop1Array(iterations: Int) {
+  for _ in 0 ..< iterations {
+    precondition(findOneOnUTF8Collection(longArrayFoFoFoFox!, needle: UInt8(ascii: "x")) != nil)
+  }
+}
+
+@inline(never)
+func runBenchRecursive3Substring(iterations: Int) {
+  for _ in 0 ..< iterations {
+    precondition(findThreeRecursive(longStringFoFoFoFox![...], needle1: "f", needle2: "o", needle3: "x") != nil)
+  }
+}
+
+@inline(never)
+func runBenchRecursive3String(iterations: Int) {
+  for _ in 0 ..< iterations {
+    precondition(findThreeRecursive(longStringFoFoFoFox!, needle1: "f", needle2: "o", needle3: "x") != nil)
+  }
+}
+
+@inline(never)
+func runBenchRecursive3ArrayOfUTF8(iterations: Int) {
+  for _ in 0 ..< iterations {
+    precondition(findThreeOnUTF8Collection(longArrayFoFoFoFox!,
```

Comment:
@eeckstein @gottesmm Is the use of `precondition` here fully equivalent to traditional `CheckResults` we use in all other benchmarks?

Reply:
Yeah, CheckResult is better. It should be slightly less overhead because it's inlinable.